src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.Rewriter.cs

```diff
@@ -78,18 +80,23 @@ private ExpressionStatementSyntax GenerateAssignment(ExpressionSyntax switchExpr
                     .WithLeadingTrivia(leadingTrivia);
             }
 
-            private StatementSyntax GenerateVariableDeclaration(ExpressionSyntax switchExpression, SyntaxTriviaList leadingTrivia)
+            private StatementSyntax GenerateVariableDeclaration(ExpressionSyntax switchExpression, SyntaxTriviaList leadingTrivia, ITypeSymbol variableSymbolTypeOpt)
             {
                 Debug.Assert(_assignmentTargetOpt is IdentifierNameSyntax);
 
+                // There is a probability that we cannot use var if the declaration type is a reference type or nullable type.
+                // In these cases, we generate the explicit type for now and decide later whether or not to use var.
+                var cannotUseVar = variableSymbolTypeOpt != null && (variableSymbolTypeOpt.IsReferenceType || variableSymbolTypeOpt.IsNullable());
```

Comment:
What does it mean if `variableSymbolTypeOpt` is null here?

Reply:
I ended up renaming the variable to `declaratorToRemoveOpt` for clarity -- if it's null, then we're not removing the original declaration of the variable, and by consequence the original issue doesn't really apply to these cases, such as below:
```C#
    int M(int i)
    {
        [||]switch (i)
        {
            case 1:
                return 0;
            case 2:
                goto default;
            default:
                return 2;
        }
    }
```
and here:
```C#
    void Test(string name)
    {
        Banana fruit = new Banana();
        switch (name)
        {
            case "apple":
                fruit = new Banana();
            break;
            case "banana":
                fruit = new OrganicBanana();
            break;
            default:
                throw new InvalidOperationException("Unknown fruit.");
        }
    }
```